crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,878 @@
+/// - Sorting `__slots__` or `__match_args__` in a class scope
+use std::borrow::Cow;
+use std::cmp::Ordering;
+
+use ruff_python_ast as ast;
+use ruff_python_codegen::Stylist;
+use ruff_python_parser::{lexer, Mode, Tok, TokenKind};
+use ruff_python_stdlib::str::is_cased_uppercase;
+use ruff_python_trivia::leading_indentation;
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextRange, TextSize};
+
+use is_macro;
+use itertools::Itertools;
+use natord;
+
+/// An enumeration of the different sorting styles
+/// currently supported for displays of string literals
+#[derive(Debug)]
+pub(super) enum SortingStyle {
+    Natural,
+    Isort,
+}
+
+impl SortingStyle {
+    pub(super) fn compare(&self, a: &str, b: &str) -> Ordering {
+        match self {
+            Self::Natural => natord::compare(a, b),
+            Self::Isort => IsortSortKey::from(a).cmp(&IsortSortKey::from(b)),
+        }
+    }
+}
+
+/// A struct to implement logic necessary to achieve
+/// an "isort-style sort".
+///
+/// An isort-style sort sorts items first according to their casing:
+/// `SCREAMING_SNAKE_CASE` names (conventionally used for global constants)
+/// come first, followed by CamelCase names (conventionally used for
+/// classes), followed by anything else. Within each category,
+/// a [natural sort](https://en.wikipedia.org/wiki/Natural_sort_order)
+/// is used to order the elements.
+struct IsortSortKey<'a> {
+    category: InferredMemberType,
+    value: &'a str,
+}
+
+impl Ord for IsortSortKey<'_> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.category
+            .cmp(&other.category)
+            .then_with(|| natord::compare(self.value, other.value))
+    }
+}
+
+impl PartialOrd for IsortSortKey<'_> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for IsortSortKey<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Eq for IsortSortKey<'_> {}
+
+impl<'a> From<&'a str> for IsortSortKey<'a> {
+    fn from(value: &'a str) -> Self {
+        Self {
+            category: InferredMemberType::of(value),
+            value,
+        }
+    }
+}
+
+impl<'a> From<&'a StringSequenceItem> for IsortSortKey<'a> {
+    fn from(item: &'a StringSequenceItem) -> Self {
+        Self::from(item.value.as_str())
+    }
+}
+
+/// Classification for the casing of an element in a
+/// sequence of literal strings.
+///
+/// This is necessary to achieve an "isort-style" sort,
+/// where elements are sorted first by category,
+/// then, within categories, are sorted according
+/// to a natural sort.
+///
+/// You'll notice that a very similar enum exists
+/// in ruff's reimplementation of isort.
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Clone, Copy)]
+enum InferredMemberType {
+    Constant,
+    Class,
+    Other,
+}
+
+impl InferredMemberType {
+    fn of(value: &str) -> Self {
```

Comment:
I would typically call this `from_value`.

Reply:
This is definitely a silly hill for me to die on, but I quite like the way `InferredMemberType::of(member)` reads like an English sentence. Happy to change it if you feel strongly, though!